build/patches/add-user-scripts.v5.patch

```diff
@@ -0,0 +1,9810 @@
++          script->set_run_location(UserScript::DOCUMENT_END);
++        else if (value == kRunAtDocumentIdleValue)
++          script->set_run_location(UserScript::DOCUMENT_IDLE);
++        else {
++          error_message = "Invalid RunAtDeclaration " + value;
++          return false;
++        }
++      }
++
++      // TODO(aa): Handle more types of metadata.
++    }
++
++    line_start = line_end + 1;
++  }
++
++  // If no patterns were specified, default to @include *. This is what
++  // Greasemonkey does.
++  if (script->globs().empty() && script->url_patterns().is_empty())
++    script->add_glob("*");
++
++  return true;
++}
++
++// static
++bool LoadUserScriptFromFile(
++    const base::FilePath& user_script_path, const GURL& original_url,
++    std::unique_ptr<UserScript>& script,
++    base::string16* error) {
++
++    std::string script_key = user_script_path.BaseName().value();
++
++    std::string content;
++    if (user_script_path.IsContentUri()) {
++      //LOG(INFO) << "---Path " << user_script_path << " is a content uri";
++
++      base::FilePath tempFilePath;
++      if( base::CreateTemporaryFile(&tempFilePath) == false ) {
++        *error = base::ASCIIToUTF16("Could not read create temp file.");
++        return false;
++      }
++
++      if( base::CopyFile(user_script_path, tempFilePath) == false ) {
++        *error = base::ASCIIToUTF16("Could not copy file to temp file.");
++        return false;
++      }
++
++      if (!base::ReadFileToString(tempFilePath, &content)) {
++        *error = base::ASCIIToUTF16("Could not read source file from temp path.");
++        return false;
++      }
++    } else {
++      if (!base::ReadFileToString(user_script_path, &content)) {
++        *error = base::ASCIIToUTF16("Could not read source file.");
++        return false;
++      }
++    }
++
++    if (!base::IsStringUTF8(content)) {
++      *error = base::ASCIIToUTF16("User script must be UTF8 encoded.");
++      return false;
++    }
++
++    std::string detailed_error;
++    if (!UserScriptLoader::ParseMetadataHeader(content, script, detailed_error)) {
++      *error = base::ASCIIToUTF16("Invalid script header. " + detailed_error);
++      return false;
++    }
++
++    // add into key the filename
++    // this value is used in ui to discriminate scripts
++    script->set_key(script_key);
++
++    script->set_match_origin_as_fallback(MatchOriginAsFallbackBehavior::kNever);
++
++    // remove unicode chars and set content into File
++    stripUnicode(content);
++    std::unique_ptr<UserScript::File> file(new UserScript::File());
++    file->set_content(content);
++    file->set_url(GURL(/*script_key*/ "script.js")); // name doesn't matter
++
++    // create SHA256 of file
++    char raw[crypto::kSHA256Length] = {0};
++    std::string key;
++    crypto::SHA256HashString(content, raw, crypto::kSHA256Length);
++    base::Base64Encode(base::StringPiece(raw, crypto::kSHA256Length), &key);
```

Comment:
that hash should be used together with the file signature and the author's public key, which should be managed in the settings.

this is the method to use
https://source.chromium.org/chromium/chromium/src/+/master:crypto/signature_verifier.cc;l=81
and here an example https://source.chromium.org/chromium/chromium/src/+/master:extensions/browser/verified_contents.cc;l=311

to calculate the hash you can also use this https://source.chromium.org/chromium/chromium/src/+/master:crypto/secure_hash.cc;bpv=1;bpt=1

To read the signature and the key you can use this
https://source.chromium.org/chromium/chromium/src/+/master:extensions/common/extension.cc;l=270

the question is what parameters are needed by openssl to produce the file with those characteristics? I must have a definitely correct file otherwise I don't know if the problem is the file or the code I'm going to write